### PR TITLE
[No QA] Build Cherry-pick flow for Expensify.cash

### PR DESCRIPTION
**Pull request by roryabraham**
**Note:** This PR creates a CP-staging workflow _only_. Not CP-production.

### Details
Creates the ability to cherry-pick a workflow to staging by merging a pull request with the `CP Staging` label.

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/155229

### Testing plan
1. Ensure that the `StagingDeployCash` is _not_ locked
1. Create a pull request with the `CP Staging` label. `OSBotify` should comment on the pull request with a comment that looks like:

    image

1. Merge the pull request with the `CP Staging` label. It should follow the normal process and be deployed to staging.
1. Lock the `StagingDeployCash`. The normal chain-of-events should happen, culminating with a staging deploy.  The `cherryPick.yml` workflow should _not_ be executed.
1. Merge a pull request with an easily identifiable change (no `CP Staging` label). Call this PR A. As per usual, it _should not_ be deployed or added to the `StagingDeployCash`. A comment should appear on the PR stating that it will be deployed later.
1. @Jag96 - Merge a pull request with an easily identifiable change with the `CP Staging` label on the PR. Call this PR B. A new version should be created and merged to main, and the `preDeploy` workflow should synchronously execute the `cherryPick.yml` workflow.
1. A staging deploy should occur, and a comment should appear on PR B stating that it was _cherry picked_ (not just deployed) to staging in the correct version.
1. Once the staging deploy completes, verify that PR A is not present on staging, but PR B is. This is where the easily-identifiable changes will come in handy.
1. Verify that PR B was added to the `StagingDeployCash`, but PR A was not.
1. @Jag96 - Use the GitHub UI to attempt to manually CP PR A to staging. The workflow should not complete. Verify that no staging deploy happens.
1. @AndrewGable or @roryabraham - Use the GitHub UI to attempt to manually CP PR A to staging. It should work and a staging deploy should occur.
1. Once the staging deploy completes, verify that PR A is present on staging.
1. Verify that PR A was added to the `StagingDeployCash`.
1. Unlock the `StagingDeployCash` and revert any temporary changes.

### CODE REVIEW

File: .github/actions/getMergeCommitForPullRequest/action.yml
Comment (AndrewGable) [inline .github/actions/getMergeCommitForPullRequest/action.yml:4]: Is this used?

**Comment by AndrewGable** [comment]
I know this is a late to the game comment, but did we explore adding the ability to add the label after the PR is merged? Or will that be able to be added in a future iteration?

**Comment by roryabraham** [comment]
> did we explore adding the ability to add the label after the PR is merged

Yeah, so this was designed with a specific vision in mind. Basically in many places we currently have `if: github.actor == 'OSBotify'`, we'll instead use an action to check that the actor is a member of a Expensify.cash Deployers GH group. That will allow certain privileged individuals to manually run/restart workflows that currently can only be run by `OSBotify`.

With that in mind, the `cherryPick.yml` workflow runs on the `workflow_dispatch` event, which means it can only be started "manually". So what we'll be able to do is manually trigger the `cherryPick.yml` workflow using the GH UI, and all you'll have to do is provide the number of the pull request you want to CP. It would look something like this:

image

File: .github/actions/getMergeCommitForPullRequest/action.yml
Comment (roryabraham) [reply .github/actions/getMergeCommitForPullRequest/action.yml:4]: Ah, yes it is. This is a pretty significant change you missed while you were gone...

Basically the `markPullRequestsAsDeployed` action was encountering GitHub rate-limiting issues because we were trying to quickly comment on many PRs in quick succession. To fix that, we [started throttling our use of the GH api](https://github.com/Expensify/Expensify.cash/pull/2769). It turned out that the changes were pretty generic, and there was no case where we specific _don't_ want throttling, so we decided to implement the throttling in the GithubUtils class itself. So now it no longer takes an `octokit` instance in its constructor, but instead [maintains a singleton octokit instance which it initializes itself](https://github.com/Expensify/Expensify.cash/blob/ed2f64bb20abc9ae37faeffc9ffca0bf53a62e57/.github/libs/GithubUtils.js#L20-L55). And now _all_ the `GithubUtils`  functions are static. However, [it still grabs the token using the actions/core library](https://github.com/Expensify/Expensify.cash/blob/ed2f64bb20abc9ae37faeffc9ffca0bf53a62e57/.github/libs/GithubUtils.js#L32), so all the actions that use it still need this input.

**Comment by AndrewGable** [comment]
Wow I had no idea that UI existed, looks great!

File: .github/workflows/cherryPick.yml
Comment (Jag96) [inline .github/workflows/cherryPick.yml:48]: nab - Just my preference, I'd remove the emojis from the PR title unless they serve a purpose

File: .github/workflows/cherryPick.yml
Comment (Jag96) [inline .github/workflows/cherryPick.yml:66]: Can we add an SO link to the PR Body that has instructions on what to do in this case if the PR isn't mergeable? Just to make it clear what the required next steps are.

**Review by Jag96 — COMMENTED** [review]
Looking good, just 2 comments

File: .github/workflows/cherryPick.yml
Comment (roryabraham) [reply .github/workflows/cherryPick.yml:48]: Emojis ftw 😄

File: .github/workflows/cherryPick.yml
Comment (roryabraham) [reply .github/workflows/cherryPick.yml:66]: Good idea!

**Comment by roryabraham** [comment]
If a PR is created with the `CP Staging`, let's add a comment like "Watch out!" and explain what will happen if it's merged.

File: .github/workflows/cherryPick.yml
Comment (roryabraham) [reply .github/workflows/cherryPick.yml:66]: Instead of a SO, how about just a comment like this:

```
This pull request has a merge conflict and could not be automatically merged.
Please manually resolve the conflicts, push your changes, and then request another reviewer to review and merge.
```

**Comment by roryabraham** [comment]
Okay, this is updated and ready for another review. Added some botify comments and opened up the `deployBlocker.yml` workflow for all members of the [Expensify.cash Deployers GH Team](https://github.com/orgs/Expensify/teams/expensify-cash-deployers)

File: .github/workflows/cherryPick.yml
Comment (Jag96) [reply .github/workflows/cherryPick.yml:79]: Perfect 👌

**Review by Jag96 — APPROVED** [review]
LGTM 👍 Let me know if you need help testing this!

File: .github/workflows/cherryPick.yml
Comment (AndrewGable) [inline .github/workflows/cherryPick.yml:17]: Can we add a test that verifies this works? Maybe @Jag96 can try to cherry pick?

**Review by AndrewGable — APPROVED** [review]

**Comment by AndrewGable** [comment]
Looks great, feel free to test without me (or if I am around I am happy to help testing!)

**Review by jasperhuangg — APPROVED** [review]
Looks good! I don't have the most knowledge about this stuff so I followed as best as I could. Feel free to merge once the test that @AndrewGable mentioned is added!

File: .github/workflows/cherryPick.yml
Comment (roryabraham) [reply .github/workflows/cherryPick.yml:17]: Good idea!

Comment (roryabraham) [reply .github/workflows/cherryPick.yml:17]: Testing steps updated 👍

**Comment by roryabraham** [comment]
Marking this as `No QA` for Applause's sake, but will begin testing this now.

**Comment by roryabraham** [comment]
- [x] StagingDeployCash is not locked
- [x] PR with CP label gets warning comment: https://github.com/Expensify/Expensify.cash/pull/3144

**Comment by roryabraham** [comment]
Fixing the last failure with [this PR](https://github.com/Expensify/Expensify.cash/pull/3151)

**Comment by roryabraham** [comment]
Adjusting the test plan a bit: we are going to:

1. skip steps 6-9 for now
1. Complete steps 10-11, except use [PR B](https://github.com/Expensify/Expensify.cash/pull/3150)
1. Complete step 7-9
1. Complete steps 6-9 again with the revert PR for [PR B](https://github.com/Expensify/Expensify.cash/pull/3150)

**Comment by roryabraham** [comment]
Okay, so @Jag96 [tried to manually CP](https://github.com/Expensify/Expensify.cash/runs/2671329834?check_suite_focus=true) [this PR](https://github.com/Expensify/Expensify.cash/pull/3150), but we ran into a permissions error where the `OSBotify` token needs higher permissions.

**Comment by roryabraham** [comment]
Okay, this testing was unfortunately a failure, and we need to step back and get some input from ring 0 folks to continue. For now, we'll:

1. Revert https://github.com/Expensify/Expensify.cash/pull/3150
1. Unlock the `StagingDeployCash`
1. Revert https://github.com/Expensify/Expensify.cash/pull/3149
1. Continue discussion over here

**Comment by roryabraham** [comment]
Continuing QA on this issue with an updated plan: since we already verified steps 1-3, we're going to restart at step 4:

4. Lock the StagingDeployCash. The normal chain-of-events should happen, culminating with a staging deploy. The cherryPick.yml workflow should not be executed.
5. Merge this PR: https://github.com/Expensify/Expensify.cash/pull/3209. As per usual, it should not be deployed or added to the StagingDeployCash. A comment should appear on the PR stating that it will be deployed later.
6. Someone other than @AndrewGable or I: merge this PR: https://github.com/Expensify/Expensify.cash/pull/3208. A new `BUILD` version should be created and merged to main, and the preDeploy workflow should synchronously execute the cherryPick.yml workflow.
7. A staging deploy should occur, and a comment should appear on https://github.com/Expensify/Expensify.cash/pull/3208 stating that it was cherry picked (not just deployed) to staging in the correct version.
8. Once the staging deploy completes, verify that https://github.com/Expensify/Expensify.cash/pull/3209 is not present on staging, but https://github.com/Expensify/Expensify.cash/pull/3208 is. This is where the easily-identifiable changes will come in handy.
9. Verify that https://github.com/Expensify/Expensify.cash/pull/3208 was added to the StagingDeployCash, but https://github.com/Expensify/Expensify.cash/pull/3209 was not.
10. Someone other than @AndrewGable or I: Use the GitHub UI to attempt to manually CP https://github.com/Expensify/Expensify.cash/pull/3209 to staging. The workflow should not complete. Verify that no staging deploy happens.
11. @AndrewGable or I: Use the GitHub UI to attempt to manually CP https://github.com/Expensify/Expensify.cash/pull/3209 to staging. It should work and a staging deploy should occur.
12. Once the staging deploy completes, verify that https://github.com/Expensify/Expensify.cash/pull/3209 is present on staging.
13. Create revert-PRs for both test PRs, give both the `CP Staging` label, and merge both.
14. Once the later staging deploy completes, verify that there are no incorrect colors on staging.
15. Close the `StagingDeployCash` to run a prod deploy. Verify that the latest staging version is deployed to prod. We'll have locked the [current `StagingDeployCash`](https://github.com/Expensify/Expensify.cash/issues/3199) and CP'd 4 PRs, so I expect that version to be `1.0.57-4`.

**Comment by roryabraham** [comment]
4. https://github.com/Expensify/Expensify.cash/runs/2696159849?check_suite_focus=true ✅ 
5. https://github.com/Expensify/Expensify.cash/runs/2696167506?check_suite_focus=true ✅ 
    - Note: we [discovered a race condition](https://github.com/Expensify/Expensify.cash/pull/3173#issuecomment-850556711) while testing this, so https://github.com/Expensify/Expensify.cash/pull/3209 _was_ deployed to staging. Should be fixed in https://github.com/Expensify/Expensify.cash/pull/3173, but for now we'll need to adjust our testing plan a bit. NBD, nothing seriously broken yet.

**Pull request merged**
